processing/src/main/java/io/druid/segment/MapVirtualColumn.java

```diff
@@ -0,0 +1,202 @@
+import io.druid.query.dimension.DefaultDimensionSpec;
+import io.druid.query.filter.DimFilterUtils;
+import io.druid.segment.data.IndexedInts;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+/**
+ */
+public class MapVirtualColumn implements VirtualColumn
+{
+  private static final byte VC_TYPE_ID = 0x00;
+
+  private final String outputName;
+  private final String keyDimension;
+  private final String valueDimension;
+
+  @JsonCreator
+  public MapVirtualColumn(
+      @JsonProperty("keyDimension") String keyDimension,
+      @JsonProperty("valueDimension") String valueDimension,
+      @JsonProperty("outputName") String outputName
+  )
+  {
+    Preconditions.checkArgument(keyDimension != null, "key dimension should not be null");
+    Preconditions.checkArgument(valueDimension != null, "value dimension should not be null");
+    Preconditions.checkArgument(outputName != null, "output name should not be null");
+
+    this.keyDimension = keyDimension;
+    this.valueDimension = valueDimension;
+    this.outputName = outputName;
+  }
+
+  @Override
+  public ObjectColumnSelector init(String dimension, ColumnSelectorFactory factory)
+  {
+    final DimensionSelector keySelector = factory.makeDimensionSelector(DefaultDimensionSpec.of(keyDimension));
+    final DimensionSelector valueSelector = factory.makeDimensionSelector(DefaultDimensionSpec.of(valueDimension));
+
+    int index = dimension.indexOf('.');
+    if (index < 0) {
+      return new ObjectColumnSelector<Map>()
+      {
+        @Override
+        public Class classOfObject()
+        {
+          return Map.class;
+        }
+
+        @Override
+        public Map get()
+        {
+          final IndexedInts keyIndices = keySelector.getRow();
+          final IndexedInts valueIndices = valueSelector.getRow();
+          if (keyIndices == null || valueIndices == null) {
+            return null;
+          }
+          final int limit = Math.min(keyIndices.size(), valueIndices.size());
+          final Map<String, String> map = Maps.newHashMapWithExpectedSize(limit);
+          for (int i = 0; i < limit; i++) {
+            map.put(
+                keySelector.lookupName(keyIndices.get(i)),
+                valueSelector.lookupName(valueIndices.get(i))
+            );
+          }
+          return map;
+        }
+      };
+    }
+
+    final int keyId = keySelector.lookupId(dimension.substring(index + 1));
+
+    return new ObjectColumnSelector<String>()
+    {
+      @Override
+      public Class classOfObject()
+      {
+        return String.class;
+      }
+
+      @Override
+      public String get()
+      {
+        final IndexedInts keyIndices = keySelector.getRow();
+        final IndexedInts valueIndices = valueSelector.getRow();
+        if (keyIndices == null || valueIndices == null) {
+          return null;
+        }
+        final int limit = Math.min(keyIndices.size(), valueIndices.size());
+        for (int i = 0; i < limit; i++) {
```

Comment:
If values of a row in a dimensions is sorted, is it possible to make use of it to become faster?